### PR TITLE
Hide password in setup-config.php

### PR DESCRIPTION
**Pull request by bgoewert**
Trac ticket: https://core.trac.wordpress.org/ticket/3534

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/js/_enqueues/admin/setup-config.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/**
```
Comment (costdev) [inline src/js/_enqueues/admin/setup-config.js]: This file could be named `password-toggle.js` for reusability on other screens in Core. What do you think?

Comment (bgoewert) [reply src/js/_enqueues/admin/setup-config.js]: Ah, missed that. Agreed. Renaming now.

File: src/js/_enqueues/admin/password-toggle.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/**
+ * Adds show/hide toggle for the setup config database password.
+ *
+ * @since 6.2.0
+ * @output wp-admin/js/setup-config.js
```
Comment (mukeshpanchal27) [inline src/js/_enqueues/admin/password-toggle.js]:
```suggestion
 * @output wp-admin/js/password-toggle.js
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Update inline comment file name.

**Comment by ironprogrammer** [comment]
@bgoewert, I like the direction this is headed, toward a shared implementation for the toggle feature 👍🏻 

Once the above feedback is integrated, I'd love to give this a go. It should allow for #2070 to be simplified as well.

File: src/js/_enqueues/admin/password-toggle.js (hunk 2 above)
Comment (bgoewert) [reply src/js/_enqueues/admin/password-toggle.js]: This has been updated with [ab58f95](https://github.com/WordPress/wordpress-develop/pull/3725/commits/ab58f95532c22d4d24af6dad9b9a01e8ace005c2). (I failed to see that you could "commit suggestion" here in GitHub)

**Comment by bgoewert** [comment]
> @bgoewert, I like the direction this is headed, toward a shared implementation for the toggle feature 👍🏻
> 
> Once the above feedback is integrated, I'd love to give this a go. It should allow for #2070 to be simplified as well.

@mukeshpanchal27's suggested change has been committed. Waiting to be reviewed again.

File: src/js/_enqueues/admin/password-toggle.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+/**
+ * Adds show/hide toggle for the setup config database password.
```
Comment (sabernhardt) [inline src/js/_enqueues/admin/password-toggle.js]: Reword this to include the possibility of the login page and/or other places.
For example: "Adds toggle functionality for password visibility buttons to switch between text and password input types."

File: src/js/_enqueues/admin/password-toggle.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * @output wp-admin/js/password-toggle.js
+ */
+
+( function () {
+	var toggle, status, input, icon, label;
+
+	toggle = document.querySelectorAll('.pwd-toggle');
+
+	toggle.forEach( function (t) {
+		t.classList.remove('hide-if-no-js');
```
Comment (sabernhardt) [inline src/js/_enqueues/admin/password-toggle.js]: Add spaces around single quotes for consistency.

File: src/wp-admin/css/forms.css [hunk 5]
```diff
@@ -550,7 +550,53 @@ fieldset label,
 }
 
 .wp-pwd {
-	margin-top: 1em;
+	display: inline-block;
```
Comment (sabernhardt) [inline src/wp-admin/css/forms.css]: This container needs to be `block` so it expands to the full width at narrower screen widths. Replacing the `span` with a `div` (as it is on the Installation page) should make the CSS unnecessary.

password input is too narrow at 600px browser width

File: src/wp-admin/css/forms.css [hunk 6]
```diff
@@ -550,7 +550,53 @@ fieldset label,
...
+	margin-top: 0;
```
Comment (sabernhardt) [inline src/wp-admin/css/forms.css]: The Profile page still should have the `1em` top margin. I suggest adding `.wp-pwd { margin-top: 0; }` to `install.css`  because the top margin should be removed from the Installation page, too. (The Login page already overrides this, setting all margin to `0` with the `.login *` selector.)

File: src/wp-admin/css/forms.css [hunk 7]
```diff
@@ -550,7 +550,53 @@ fieldset label,
...
+	position: relative;
+}
+
+.wp-pwd input {
```
Comment (sabernhardt) [inline src/wp-admin/css/forms.css]: Remove the `.wp-pwd input` styles from `forms.css`. The `#pwd` direction is already [set to LTR](https://github.com/WordPress/wordpress-develop/blob/53063c633306997e2502a7ea90c90b58236f7183/src/wp-admin/css/install.css#L210), and the side padding here would also apply to the password input on the Profile page.

padding affects Profile page

File: src/wp-admin/css/install.css [hunk 8]
```diff
@@ -151,6 +151,10 @@ textarea {
 	padding: 3px 5px;
 }
 
+.form-table .wp-pwd input {
+	padding-right: 40px;
```
Comment (sabernhardt) [inline src/wp-admin/css/install.css]:
```
#pwd {
	padding-right: 2.5rem;
}
```

I prefer the `2.5rem` value instead of pixels, as it is on the Login page.

The selector (either the ID or a new class) needs to apply only on the `setup-config` page. The `.form-table .wp-pwd input` selector also affects the Installation page.

padding on Install page

File: src/wp-admin/setup-config.php [hunk 9]
```diff
@@ -223,7 +223,14 @@ function setup_config_display_header( $body_classes = array() ) {
 		</tr>
 		<tr>
 			<th scope="row"><label for="pwd"><?php _e( 'Password' ); ?></label></th>
-			<td><input name="pwd" id="pwd" type="text" aria-describedby="pwd-desc" size="25" placeholder="<?php echo htmlspecialchars( _x( 'password', 'example password' ), ENT_QUOTES ); ?>" autocomplete="off" /></td>
+			<td>
+				<span class="wp-pwd">
+					<input name="pwd" id="pwd" type="password" class="regular-text" data-reveal="1" data-pw="<?php echo esc_attr( get_option( '' ) ); ?>" aria-describedby="pwd-desc" size="25" placeholder="<?php echo htmlspecialchars( _x( 'password', 'example password' ), ENT_QUOTES ); ?>" autocomplete="off" />
```
Comment (sabernhardt) [inline src/wp-admin/setup-config.php]: Leave the `data-pw` attribute empty.

File: src/wp-admin/setup-config.php [hunk 10]
```diff
@@ -223,7 +223,14 @@ function setup_config_display_header( $body_classes = array() ) {
...
+					<button type="button" class="button pwd-toggle" data-toggle="0" data-start-masked="1" aria-label="<?php esc_attr_e( 'Hide password' ); ?>">
```
Comment (sabernhardt) [inline src/wp-admin/setup-config.php]: 1. The toggle never changes the ARIA label, so it says "Hide password" even when the password is hidden.
2. This button is unusable without JS, so it needs the `hide-if-no-js` class.

**Review by sabernhardt — CHANGES_REQUESTED** [review]
Thanks for creating a script to try making this more consistent. If we keep the icon-only button, reusing the script for the Login page could help reduce resources there (ticket:57157).

However, while the icon-only button is consistent with the Login page, it is not consistent with other visibility toggle buttons in `.form-table` elements at larger screen sizes (on Profile and Installation pages). Andrea acknowledged that icon-only controls are an [accessibility anti-pattern](https://core.trac.wordpress.org/ticket/42888#comment:51) before committing [r46256](https://core.trac.wordpress.org/changeset/46256) as a "self-contained" improvement that ideally would have a different design later.

Also, the icon-only button on `setup-config` has the overlap with some password managers, as seen on the Login page (ticket:48222).

1. In Edge, the LastPass icon appears behind the button on `setup-config`, so the toggle button is still clickable.

   setup-config with LastPass in Edge

2. In Chrome (Windows), the Norton Password Manager icon covers the button at a higher `z-index`:

   setup-config with Norton Password Manager in Chrome

I marked some changes to this PR in case we continue in this direction.

File: src/wp-admin/setup-config.php (hunk 9 above)
Comment (bgoewert) [reply src/wp-admin/setup-config.php]: I'm not sure this `data-pw` even needs to be there, does it? I'll confess that some of these may have been copied from 2070 or 2071 and I just didn't review it close enough. From what I can tell, it can be removed from `setup-config.php`.

File: src/wp-admin/css/forms.css (hunk 5 above)
Comment (bgoewert) [reply src/wp-admin/css/forms.css]: I've fixed the spanning issue for narrower screen sizes. However, I also noticed the password manager icons are overlapping the toggle button. When I have more time I'll attempt to address this as well. Leaving this open until then.

Screenshot_2023-01-05_22-49

File: src/wp-admin/setup-config.php (hunk 9 above)
Comment (bgoewert) [reply src/wp-admin/setup-config.php]: I removed the `data-pw`, as it seems unused. #2071 also did not include this.

File: src/wp-admin/setup-config.php (hunk 10 above)
Comment (bgoewert) [reply src/wp-admin/setup-config.php]: 1. Added ability to change `aria-label` with language support. I was unsure how best to support different translations and ended up added a data attribute to store both.
2. Added the `hide-if-no-js` class back (I may or may not have removed it).

File: src/wp-admin/css/forms.css (hunk 5 above)
Comment (bgoewert) [reply src/wp-admin/css/forms.css]: In an attempt to fix the overlap issue I commented on, as well as the incorrect lang support [I added here](https://github.com/WordPress/wordpress-develop/pull/3725/commits/31210f927c70885775c7b80df56fba0ca54ec3a9), I noticed that the profile page was hiding the pw toggle text on narrower screens. So I added `<span class="text"><?php _e( 'Show' ); ?></span>` to test. Unfortunately, this creates white space that is not taken up by the pw field or pw manager icons and the button is still focusable. It may be better to hide the button entirely and re-add the 10px padding that is added to input fields [here](https://core.trac.wordpress.org/browser/tags/6.1/src/wp-admin/css/install.css#L283) for narrower screens.

Screenshot_2023-01-05_23-56

Thoughts @sabernhardt?

**Comment by bgoewert** [comment]
Thank you @sabernhardt for the thorough review. I tried to address all your comments. Another change I added towards a standardized `password-toggle.js` is [this fix](https://github.com/WordPress/wordpress-develop/pull/3725/commits/605a8d61472c98177828a961118ea83d31688b49) for performing the toggle to the correct parent input.

> However, while the icon-only button is consistent with the Login page, it is not consistent with other visibility toggle buttons in .form-table elements at larger screen sizes (on Profile and Installation pages). Andrea acknowledged that icon-only controls are an [accessibility anti-pattern](https://core.trac.wordpress.org/ticket/42888#comment:51) before committing [r46256](https://core.trac.wordpress.org/changeset/46256) as a "self-contained" improvement that ideally would have a different design later.

I noticed the profile page has an icon and text. If it helps with accessibility I can change the icon to text, or add an icon and text, to maintain consistency with the other toggles. Would it be preferable to still keep the button as "part of" the password input? Or do we go back to separating the button from the input? Do we then at some point also change the login page to follow suit?

> Also, the icon-only button on setup-config has the overlap with some password managers, as seen on the Login page (ticket:48222).

Separating the button from the input would address this.

File: src/wp-admin/css/forms.css (hunk 5 above)
Comment (bgoewert) [reply src/wp-admin/css/forms.css]: This was fixed in PR #3827 by separating the button from the input. Technically this is unresolved in this PR so I will leave it as such.

**Comment by bgoewert** [comment]
Closed in [r56008](https://core.trac.wordpress.org/changeset/56008)
